pkg/framework/plugins/nodeutilization/nodeutilization.go

```diff
@@ -523,3 +425,75 @@ func classifyPods(pods []*v1.Pod, filter func(pod *v1.Pod) bool) ([]*v1.Pod, []*
 
 	return nonRemovablePods, removablePods
 }
+
+// AssessNodesUsagesAndThresholds converts the raw usage data into percentage.
+// In case we need to calculate deviations from the average this function does
+// so as well. Thresholds are provided in percentage but they slightly differ
+// in behavior: if no deviation is enabled they are interpreted as is but with
+// deviations enabled they are considered deviations from the node's average.
+// Returns the usage (pct) and the thresholds (pct) for each node.
+func AssessNodesUsagesAndThresholds(
+	rawUsages, rawCapacities map[string]api.ReferencedResourceList,
+	lowSpan, highSpan api.ResourceThresholds,
+	calculateDeviation bool,
+) (map[string]api.ResourceThresholds, map[string][]api.ResourceThresholds) {
+	// first we normalize the node usage from the raw data (Mi, Gi, etc)
+	// into api.Percentage values.
+	usage := normalizer.Normalize(
+		rawUsages, rawCapacities, ResourceUsageNormalizer,
+	)
+
+	// if we are not taking the average and applying deviations to it we
+	// can simply replicate the same threshold across all nodes and return.
+	if !calculateDeviation {
+		thresholds := normalizer.Replicate(
+			slices.Collect(maps.Keys(rawUsages)),
+			[]api.ResourceThresholds{lowSpan, highSpan},
+		)
+		return usage, thresholds
+	}
+
+	// calculate the average usage and then deviate it according to the
+	// user provided thresholds. We also ensure that the value after the
+	// deviation is at least 1%. this call also replicates the thresholds
+	// across all nodes.
+	thresholds := normalizer.Replicate(
+		slices.Collect(maps.Keys(rawUsages)),
+		normalizer.Min(
+			normalizer.Deviate(
+				normalizer.Average(usage),
+				[]api.ResourceThresholds{
+					lowSpan.Negative(),
+					highSpan,
+				},
+			),
+			1,
+		),
+	)
+
+	return usage, thresholds
+}
+
+// ResourceUsageNormalizer is an implementation of a Normalizer that converts a
+// set of resource usages and totals into percentage. This function operates on
+// Quantity Value() for all the resources except CPU, where it uses
+// MilliValue().
+func ResourceUsageNormalizer(usages, totals api.ReferencedResourceList) api.ResourceThresholds {
+	result := api.ResourceThresholds{}
+	for rname, value := range usages {
+		total, ok := totals[rname]
+		if !ok || value == nil || total == nil {
```

Comment:
We should panic when `usages` and `totals` lists do not have identical keys. Anyone calling `ResourceUsageNormalizer` needs to make sure the incoming lists are valid.

Reply:
We do have some tests where we expect some nodes to do not have resource capacities set, see [here](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/framework/plugins/nodeutilization/highnodeutilization_test.go#L263). Should we start panic on these scenarios ? [This](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/framework/plugins/nodeutilization/highnodeutilization_test.go#L309) test even mentions that the extra resources won't exist in some nodes.

Reply:
If a node does not have the corresponding capacity for a given resourcename set, invoker needs to make sure the capacity is extended with resourcename with a zero quantity.

Reply:
Yes. This can be skipped for now and fixed later.

Reply:
Done.